Add a wildcard [type:xxx] modifier

krader1961 opened the pull request:
This allows filtering the wildcard expansion by path type. For example,
whether the path is a directory, regular file, or something else. So
instead of `find . -type f` you can now do `**[type:file]`.

Resolves #1015

xiaq commented:
Thanks for tackling this. There are some open design questions that I don't have good answers for, and I suggest we start with a modest feature set and iterate on this.

Ultimately, I'd like Elvish to have a small, general-purpose query syntax with support for NOT, AND and OR relations, and the glob modifiers syntax should be just an example of it. The same syntax can be used, for example, in the navigation mode to filter files based on their properties (which should use exactly the same syntax as glob modifiers), and in command history listing to filter out history items (when they have metadata such as working directory, duration, etc.).

I haven't reviewed the code in detail, so some design comments:

* Juxtaposition of filters strongly suggests AND: this is how most search engine's query languages work. I'd prefer to start simply forbidding multiple `type:` in this PR before we agree on the design of how to do AND and OR.

* The abbreviated form for `type:dir`, `type:d`, isn't actually very short. I'd just use `/`; which is easy to remember and consistent with `ls -F`. There is no suffix for regular files but I'd copy `.` from zsh's list of single-letter glob modifiers (see section 14.8.7 [here](http://zsh.sourceforge.net/Doc/Release/Expansion.html)). In fact, Elvish's glob modifier feature is inspired by Zsh's, except that it is done right ;-)

* There is no well-defined "other" category; "not a regular file and not a directory" is an arbitrary definition. I suggest dropping it from the PR, and maybe add all the special device classes etc later. Also, what is currently "other" can be expressed in terms of the NOT syntax, when we have that designed.

xiaq commented:
To see why the AND semantics for juxtaposition is useful and intuitive, consider that we may want to add filters for owners and permissions. It's intuitive to write `**[type:dir owner:xiaq]`, and the underlying relationship is AND.

xiaq commented:
Some more comments:

* I prefer `type:regular` to `type:file`. Nominally directories are also files, so `type:file` is not clear. "Regular file" is not 100% clear either but definitely an improvement.

* For the short form: not all filters need to have a short form, only when they are used a lot and the short form is very intuitive. So I'd suggest starting with only `/` for `type:dir`, and live without a short form for `type:regular`.

xiaq commented:
For the general purpose query syntax, the [search query language](https://developers.google.com/issue-tracker/concepts/search-query-language) of Google's issue tracker is a good model, but unfortunately much of it won't work directly with Elvish due to syntax clashes. Elvish's query syntax should play well with the language syntax.

xiaq commented:
Oh, and perhaps more notably, there is also the [syntax](https://developers.google.com/code-search/reference) of Google's codesearch, which is very similar.

krader1961 commented:
Gah! Please don't implement a query DSL just for use in wildcard expansion. Better to allow a lambda as a callback that takes two args: the path and the `os.FileInfo` object. It can then use Elvish, along with the commands from the proposed `path:` module to do arbitrary filtering.

> I prefer type:regular to type:file

I'm going to disagree. Not only is "regular" longer it is less clear. Most UNIX users would be surprised if you told them directories are just special files. But everyone, including geeks like us, understand

> I'd prefer to start with simply forbidding multiple type:

Works for me. In practice I would not expect anyone to ever use more than one `type:` modifier. And as you note the implicit "or" is inconsistent with how the modifiers otherwise work.

> The abbreviated form for type:dir, type:d, isn't actually very short. I'd just use /; which is easy to remember and consistent with ls -F. There is no suffix for regular files but I'd copy . from zsh's list of single-letter glob modifiers.

I am **not** a fan of zsh single character glob qualifiers. And I say that despite having been a fan of Perl, a language some have described as serial line noise, for many years. Way too magical for my taste. Worse, they require memorizing an otherwise unimportant details (the list of magic single char qualifiers) or having to perform a difficult search to find out what `[/]` means. Clarity is more important in this case than avoiding typing seven characters.

krader1961 commented:
I removed `[type:other]` and support for more than one type modifier. I'd like to leave introduction of single-char modifiers such as `[/]` until there has been more discussion about the merits of going down that path. Please take another look.

xiaq commented:
A generic functional API and DSL are not conflicting goals. Introducing `/` as a shorthand for `type:dir` does not mean copying Zsh's qualifiers wholesale.

krader1961 commented:
> Introducing / as a shorthand for type:dir does not mean copying Zsh's qualifiers wholesale.

Yes, but my intuition, and decades of experience, is that restricting the expansion to directories is much less common than restricting it to regular files. Consider how often you use a wildcard to generate a list of files to be passed to commands like `grep`. Versus how often you use a wildcard to generate a list of directories. The reason I have a `ff` function is that 99% of the time I want to enumerate all the regular files at or below the CWD.

If either of those two cases deserve a single-char shorthand it is the regular file case. So I'm willing to introduce `[/]` as an alias for `[type:dir]` only if we can agree on a single char alias for `[type:file]`. AFAICT, bash is particularly brain-dead in this regard as it doesn't support restricting matches to regular files and and has a rather obtuse mechanism for matching directories:

```
              *      Matches any string, including the null  string.   When  the  globstar
                     shell  option  is enabled, and * is used in a pathname expansion con-
                     text, two adjacent *s used as a single pattern will match  all  files
                     and zero or more directories and subdirectories.  If followed by a /,
                     two adjacent *s will match only directories and subdirectories.
```

Is there some reason we can't just support `[d]` and `[f]`? AFAICT those do not conflict with any existing wildcard modifiers an if we're going to support single-char modifiers I don't see any reason to prefer `[/]` over `[d]`.

xiaq commented:
* My experience and intuition are different from you, but I am convinced that restricting wildcards to regular files is also a frequent use case.

* I don't feel particularly strong about `/` as a shorthand for `type:dir`, so I'm fine with leaving it out. It could be useful for other purposes, and having to spell out `type:dir` is probably for the best.

* I remain unconvinced about using "file" to mean *regular* file. I know it's a strong tradition used by both `test(1)` and `find(1)`, but it's worth making a break here.

  It's not so problematic when it comes to the file/directory distinction, but it *is* problematic when it comes to things such as device files. The name "file" makes it really unclear whether it includes device files or not. On the other hand, "regular file" is well-defined in the [C API](https://www.man7.org/linux/man-pages/man2/stat.2.html), in the [Go API](https://godoc.org/os#FileMode.IsRegular), and in other language bindings such as [Python](https://docs.python.org/3/library/stat.html#stat.S_ISREG), where it does not include device files.

* I acknowledge `type:d` and `type:f` as as a nod `find(1)`'s `-type d` and `-type f`, but remain unconvinced about the need for them.

I suggest starting with only `type:dir` and `type:regular` and nothing more for the sake of this PR.

krader1961 commented:
PTAL. I changed "type:file" to "type:regular" and also introduced `[d]` and `[r]` as shorthand for the long forms.

xiaq commented on pkg/eval/glob.go:
```diff
@@ -15,12 +16,23 @@ import (
 
 type GlobFlag uint
 
+const (
+	typeDir     = "dir"
+	typeRegular = "regular"
+)
```
Remove

xiaq commented on pkg/eval/glob.go:
```diff
@@ -75,6 +89,24 @@ func (gp GlobPattern) Index(k interface{}) (interface{}, error) {
 		gp.Segments[len(gp.Segments)-1] = glob.Wild{
 			lastSeg.Type, true, lastSeg.Matchers,
 		}
+	case modifier == "d" || modifier == "r" || strings.HasPrefix(modifier, "type:"):
```
`type:dir` and `type:regular` and nothing more; remove "d" and "r".

krader1961 replied:
Okay. Done. That makes this more painful to use interactively but presumably you would prefer not to institute these shorthands until more experience using this feature is gained and there has been more discussion about whether shorthands like `/` for `type:dir` is preferable to `d` or some other shorthand.

xiaq commented on pkg/eval/glob.go:
```diff
@@ -75,6 +89,24 @@ func (gp GlobPattern) Index(k interface{}) (interface{}, error) {
 			lastSeg.Type, true, lastSeg.Matchers,
 		}
+	case modifier == "d" || modifier == "r" || strings.HasPrefix(modifier, "type:"):
+		// This handles cases such as `**x*[r][d]`.
```
Change this comment to use the longhand form, or just remove it as it's quite self-evident what this branch does.

xiaq commented on pkg/eval/glob.go:
```diff
@@ -75,6 +89,24 @@ func (gp GlobPattern) Index(k interface{}) (interface{}, error) {
+			return nil, ErrMultipleTypeModifiers
+		}
+		switch {
+		case modifier == "d":
+			gp.TypeCb = typeMap[typeDir]
+		case modifier == "r":
+			gp.TypeCb = typeMap[typeRegular]
```
Remove

xiaq commented on pkg/eval/glob.go:
```diff
@@ -75,6 +89,24 @@ func (gp GlobPattern) Index(k interface{}) (interface{}, error) {
+		case modifier == "r":
+			gp.TypeCb = typeMap[typeRegular]
+		default:
+			ptype := modifier[len("type:"):]
```
I know you can't use `type` since it's a keyword, but not clear what `ptype` means; `typeName` is better. Or just don't use a variable here, it is only used in the next statement.

krader1961 replied:
Renamed. Note: I prefer a temp var in this situation because it's much clearer to a human and shouldn't be any slower.

xiaq commented on pkg/eval/glob.go:
```diff
@@ -15,12 +16,23 @@ import (
+	typeRegular = "regular"
+)
+
+var typeMap = map[string]func(os.FileMode) bool{
```
`typeCbMap` would be a bit clearer.

xiaq commented on pkg/eval/glob.go:
```diff
@@ -123,6 +155,11 @@ func (gp GlobPattern) Concat(v interface{}) (interface{}, error) {
 		gp.append(rhs.Segments[0])
 		gp.Flags |= rhs.Flags
 		gp.Buts = append(gp.Buts, rhs.Buts...)
+		// This handles cases such as `**[r]x*[d]`.
```
Change this comment to use the longhand form.

xiaq commented on pkg/eval/glob.go:
```diff
@@ -123,6 +155,11 @@ func (gp GlobPattern) Concat(v interface{}) (interface{}, error) {
+		if gp.TypeCb != nil && rhs.TypeCb != nil {
+			return nil, ErrMultipleTypeModifiers
+		}
+		gp.TypeCb = rhs.TypeCb
```
This should only be done if `rhs.TypeCb != nil`.

xiaq commented on pkg/eval/glob.go:
```diff
@@ -208,15 +246,24 @@ func doGlob(gp GlobPattern, abort <-chan struct{}) ([]interface{}, error) {
+			return true
+		}
+
+		if gp.TypeCb == nil { // no file type restrictions, common case
+			vs = append(vs, pathInfo.Path)
+		} else {
+			if (gp.TypeCb)(pathInfo.Info.Mode()) {
+				vs = append(vs, pathInfo.Path)
+			}
```
This can be written a bit more succinctly as:

```go
if gp.TypeCb == nil || gp.TypeCb(pathInfo.Info.Mode()) {
    vs = append(vs, pathInfo.Path)
}
```

xiaq commented on pkg/eval/glob_test.go:
```diff
@@ -88,3 +88,23 @@ func TestGlob_But(t *testing.T) {
+	Test(t,
+		That("put **[d]").Puts("b/c", "b", "d1", "d2"),
+		That("put **[type:dir]").Puts("b/c", "b", "d1", "d2"),
+		That("put **[r]f*").Puts("d1/f1", "d2/fm", "foo"),
```
This test doesn't really test the regular modifier since only regular files have `f` in their names, meaning that it fails to capture the bug I mentioned earlier in the review (where the rhs's nil TypeCb overrides the non-nil TypeCb).

xiaq commented on pkg/glob/glob.go:
```diff
@@ -10,13 +10,18 @@ import (
 
 // TODO: Use native path separators instead of always using /.
 
+type PathInfo struct {
```
Exported symbols must have godoc comments.

krader1961 replied:
FYI, The existing code violates this best practice. A lot. I installed `golint` a few weeks ago. Simply doing `golint pkg/eval` identifies 24 such instances :-)

xiaq replied:
Yes, but I unapologetically hold new code to a higher standard :)

xiaq commented on pkg/glob/glob.go:
```diff
@@ -10,13 +10,18 @@ import (
 // TODO: Use native path separators instead of always using /.
 
+type PathInfo struct {
+	Path string
```
A reader might question why we can't just use `Info.Name()` here, so document the reason - `Path` is needed to generate paths consistent with the original glob pattern.

xiaq commented on pkg/glob/glob.go:
```diff
@@ -63,11 +68,14 @@ func glob(segs []Segment, dir string, cb func(string) bool) bool {
 
 	if len(segs) == 0 {
-		return cb(dir)
+		if info, err := os.Stat(dir); err == nil {
```
You can save the `FileInfo` the code on line 65 obtained earlier and avoid another `Stat`.

krader1961 replied:
I have no idea what you're suggesting. If `len(segs) == 0` was true when this function was entered then the preceding `os.Stat()` would not have been run.

xiaq replied:
`len(segs) == 0` will only become true if the loop has been run, which shortens `segs` (`segs = segs[2:]`), and this function is never called with an empty `segs`.

OTOH, it's not easy to reason all that. The code as it is now is easier to reason about.

xiaq commented on pkg/glob/glob.go:
```diff
@@ -63,11 +68,14 @@ func glob(segs []Segment, dir string, cb func(string) bool) bool {
 		path := dir + segs[0].(Literal).Data
-		if _, err := os.Stat(path); err == nil {
-			return cb(path)
+		if info, err := os.Stat(path); err == nil {
```
This should use `Lstat` to avoid dropping broken symlinks.

This will also drop files that you can't stat. It is a sound strategy to handle the error, since that indicates that the filesystem is changing under our feet and there isn't much we can do.

However, it's worth calling that out as an conscious choice in a comment, since this may raise a red flag to a reader as not handling the error properly.

krader1961 replied:
Since the existing, and surrounding, code uses `Stat` I'd prefer to leave that unchanged with respect to implementing this feature. If this should be changed to `Lstat` there probably needs to be a careful review of the other paths that call `Stat` and appropriate comments added as to why one or the other is used.

Similarly, when `Readdir` fails you have the comment "// XXX Silently drop the error". But that same comment isn't present in the `Stat` error cases. Since adding such comments are orthogonal to this change I'd prefer they be done in a separate change.

xiaq commented on pkg/glob/glob.go:
```diff
@@ -142,7 +150,12 @@ func glob(segs []Segment, dir string, cb func(string) bool) bool {
 		if matchElement(segs, name) {
-			if !cb(dir + name) {
+			name = dir + name
+			info, err := os.Stat(name)
+			if err != nil {
+				return true
+			}
```
You already have a `FileInfo`; this is redundant.

krader1961 replied:
I have no idea what you're telling me. There is no usable `FileInfo` object at this point in the code.

xiaq commented on pkg/glob/glob.go:
```diff
@@ -142,7 +150,12 @@ func glob(segs []Segment, dir string, cb func(string) bool) bool {
 		name := info.Name()
 		if matchElement(segs, name) {
-			if !cb(dir + name) {
+			name = dir + name
```
Do not modify `name` in place; it gets confusing.

krader1961 replied:
Agreed. Changed to `dirname`.

xiaq commented on website/ref/language.md:
```diff
@@ -949,19 +962,24 @@ pattern: `put */*.cpp[nomatch-ok]` unfortunately does not do the correct thing.
-    -   Multiple matchers, they are OR'ed. For instance, ?[set:aeoiu][digit]
-        matches `aeoiu` and digits.
+-   Local matchers combined in the same modifier, such as `?[set:aeoiu digit]`,
+    are OR'ed but in a manner something like a SQL outer join. This results in
```
This behavior is more easily and unambiguously explained by saying that `?[set:aeoiu digit]` expands to `?[set:aeiou] ?[digit]`.

krader1961 replied:
Your example does not match the behavior I'm seeing. Either standalone or bracketed by `*` wildcards. For example:

```
 macpro  ~/tmp  > echo ?[set:aeoiu digit]
1 a 1 a
 macpro  ~/tmp  > echo ?[set:aeoiu] ?[digit]
a 1
 macpro  ~/tmp  > echo *?[set:aeoiu digit]*
1 :3 a a1 b2 d:4 1 :3 a a1 b2 d:4
 macpro  ~/tmp  > echo *?[set:aeoiu]?[digit]*
a1
```

Frankly, I think combining local modifiers in this fashion should not be allowed. At least not until it has a behavior that can be explained to an average shell user. So I'm simply going to modify the comment to say "do not use".

xiaq commented on website/ref/language.md:
```diff
@@ -920,11 +922,22 @@ wildcard:
+    -   `regular` will match if the path is a regular file.
 
-On the other hand, you must add it after a wildcard, instead of after the entire
+-   `d` is shorthand for `type:dir`.
+
+-   `r` is shorthand for `type:regular`.
```
Remove.

xiaq commented:
Hi! Sorry for nagging, but do you intend to act on the feedback soon?

If you don't have time to work on this now, I can merge it and do the changes myself.

krader1961 commented:
PTAL. This addresses all comments other than a) the suggestion to replace an existing `os.Stat` with `os.Lstat`, and b) using the return value of an earlier `os.Stat` since, AFAICT, we can't count on the previous `os.Stat` to have been executed when its output is needed later in the code.

krader1961 commented:
Regarding the documentation of multiple matchers in the same modifier... the current P.R. says to avoid using that form because the behavior is not easy to describe, let alone intuitive. See https://github.com/elves/elvish/issues/1019#issuecomment-635095672 for an example why.

xiaq commented on pkg/eval/glob.go:
```diff
@@ -208,15 +238,20 @@ func doGlob(gp GlobPattern, abort <-chan struct{}) ([]interface{}, error) {
+			return true
+		}
+
+		if gp.TypeCb == nil || (gp.TypeCb)(pathInfo.Info.Mode()) {
```
Unnecessary parentheses around `gp.TypeCb`.

xiaq commented on website/ref/language.md:
```diff
@@ -920,11 +922,17 @@ wildcard:
 
-On the other hand, you must add it after a wildcard, instead of after the entire
+    -   `regular` will match if the path is a regular file.
+
```
Removed the first line of this paragraph by mistake?
